Implement Vertical GUI (Portrait Mode)

radry opened the pull request:
### Requirements

**Currently vertical GUI is only implemented for TFT24.** 
Adding all other screens is easy since the base code is all the same but requires an actual screen to test it (which I don't have).
Only screen and controller specific files have to be changed/added.
How to do this is described below.

**Compile:**
To use TFT24 in portrait mode you need to compile for "BIGTREE_TFT24_V1.1_P" (Default enviroment is set to this in this commit).

**Config.ini:**
set "rotated_ui" to 2 or 3

**Theme/Icons:**
logo.bmp must have a portrait resolution (for TFT24 that's 240x320 instead of 320x240). Do not simply rotate it, the logo must still stay the same orientation.

### Description

This is a first implementation of a vertical GUI (Portrait Mode).

**How is it implemented:** 
My way of approaching this is as follows:

- Treat the vertical orientation as if it's its own screen. That means to use a screen in vertical orientation, it needs to be compiled for portrait mode. It can then only be used in portrait mode. To use landscape mode you just use the regular .bin and reset the settings.

- Most code that handles vertical GUI is filtered by `#ifdef PORTRAIT` and thus only compiled when target is a Portrait mode screen (PORTRAIT is defined in Vartiants/pins_TFTxxxx which has its own version for vertical orientation). 
This keeps the resulting .bin small. 

### How it looks
**UPDATE 06.12.2021**

**Status Screen**
![Status_new](https://user-images.githubusercontent.com/8878118/144753303-2a156bb8-ba90-43a2-ab11-88f2421dc5f0.jpg)
**Heat Menu**
![heat_new](https://user-images.githubusercontent.com/8878118/144753310-4f733ab6-05d7-4a60-b74f-d9d30cf02807.jpg)
**Main Menu**
![main_new](https://user-images.githubusercontent.com/8878118/144754572-1319bd46-bc6f-4799-922b-a6b7fcd386e2.jpg)
**Move**
![move_new](https://user-images.githubusercontent.com/8878118/144754578-498ba8da-a1a0-4feb-b6fc-b49c6fed4d18.jpg)
**Level**
![level_new](https://user-images.githubusercontent.com/8878118/144754579-f43d468c-8864-44d6-93a1-7a12cacefcd0.jpg)
**Printing**
![printing_new_1](https://user-images.githubusercontent.com/8878118/144910010-220d159c-b2c3-43cd-a246-27171a65f000.jpg)
**List Layout**
![ListMenu](https://user-images.githubusercontent.com/8878118/143779243-f3cd0445-e68b-4f47-87c4-35817700f804.jpg)

### How to add portrait mode to other screens

_Coming Soon_

### Benefits

Adds vertical GUI

### Related Issues

#524 
#1830 
#1170

digant73 commented:
as a personal suggestion, I would bind KEY_INFOBOX area as you made in Status Screen menu (so centered at the bottom). That means the temp info in the Heat menu will be centered at the bottom of the screen.
In the Printing menu I would swap the "stop" button with "more" or "babystep". Also I would move the progress bar below the "speed" icon (so just above the "pause/resume", babystep (or "more") and "stop")

radry commented:
@digant73 Thank you for your suggestion.  

Placing the Temperature in the Heat menu (and similars) on the bottom is possible and I agree it would look cleaner but it will put the "back" button in a different location than the main menu (and most other menus). The benefit of the current implementation is, that the back button is always in the same position, regardless of the menu. 
For example you can hit "back" in the heat menu and afterwards hit "back" in the main menu in the same location. Personally I prefer the possibility to hit "back" quickly in succession without having to switch finger position every time.
I think this will lower the risk of accidently pressing a wrong button. 

This is surely a point worth discussing.

digant73 commented on TFT/src/User/API/menu.c:
```diff
@@ -184,21 +370,34 @@ const GUI_RECT rect_of_keyPS_end[] = {
 const GUI_RECT rect_of_keyListView[ITEM_PER_PAGE + 1] = {
-  // 8 icons area
+  #ifdef PORTRAIT
+  // list area VERTICAL GUI
```
four spaces, not two

digant73 commented on TFT/src/User/API/menu.c:
```diff
@@ -184,21 +370,34 @@ const GUI_RECT rect_of_keyPS_end[] = {
+  {START_X+1.5*LIST_ICON_WIDTH,  5*LISTITEM_HEIGHT+ICON_START_Y,  START_X+2.5*LIST_ICON_WIDTH,  5*LISTITEM_HEIGHT+ICON_START_Y+1*LIST_ICON_HEIGHT},
+  {LCD_WIDTH-START_X-1*LIST_ICON_WIDTH,  5*LISTITEM_HEIGHT+ICON_START_Y,  LCD_WIDTH-START_X,  5*LISTITEM_HEIGHT+ICON_START_Y+1*LIST_ICON_HEIGHT},
+  #else 
+  // list area LANDSCAPE GUI
```
four spaces, not two

digant73 commented on TFT/src/User/API/menu.c:
```diff
@@ -265,8 +464,24 @@ void menuClearGaps(void)
+//clear gaps all other screents except TFT70
 void menuClearGaps(void)
 {
+#ifdef PORTRAIT
```
you should indent this #if, #else block

digant73 commented:
@radry just a brief review.
Also I would consider to move the info area always on top (in all menus). The back button will be moved at the bottom (last row, third column).
That means for example, the Heat menu could be:
row 1: Temp info (infobox)
row 2: -, unit, +
row 3: nozzle, stop, back

radry commented:
I reworked the UI and also implemented @digant73 suggestion:

- The heat menu (and similar) has the temperature on the top.
- Thus I also changed the Status Menu to use such a layout.
- The main menu (and sub menus) have the bottom row icons now aligned left/right. This is done to have the back button always on the bottom right (same as heat menu).

HOWEVER:
Some sub menus like "move" and "level" should be reworked because the order of the icons is still not tidy. They work fine in the current state but are not pretty.
This must be done in each sub menus .c file by adjusting the order (of drawing and the key query) when compiled for portrait.
I don't know yet how many menus are affected. It will take some time to test them all.

How is this? If you like it I will push it to the PR after cleaning up the code as mentioned by @digant73
Please excuse the blurry pics.
**Status Screen**
![Status_new](https://user-images.githubusercontent.com/8878118/144753303-2a156bb8-ba90-43a2-ab11-88f2421dc5f0.jpg)
**Heat Menu**
![heat_new](https://user-images.githubusercontent.com/8878118/144753310-4f733ab6-05d7-4a60-b74f-d9d30cf02807.jpg)
**Main Menu**
![main_new](https://user-images.githubusercontent.com/8878118/144753305-998f122c-eebb-4d87-b3ba-20f2f286ceae.jpg)

**Current state of "Move"**
![move_new](https://user-images.githubusercontent.com/8878118/144753313-a3d1baca-ae2b-4eb7-a314-d4b2910370a6.jpg)
**Current state of "Level"**
![level_new](https://user-images.githubusercontent.com/8878118/144753314-32d68a3e-48f0-4a26-8cea-22dc2dda1978.jpg)

labotecno commented:
nice ! i like it

digant73 commented:
@radry The menus with the infobox area (e.g. Heat, Status Screen etc...) are now OK.
For the menus with no infobox (e.g. Main, Move, Leveling etc...) it should be enough to simply remap icon 4 on the standard layout (non portrait mode) to icon 7 on portrait mode. That means you will have:

Main menu:
row 1: Heat/Fan, Movement, Un(Load)
row 2: Terminal, Custom, Settings
row 3: Em. Stop, , Back

Move menu:
row 1: Z-, Y+, Z+
row 2: X-, Y-, X+
row 3: 0.1mm, , Back

Level menu:
row 1: Point 4, Point 3, Distance
row 2: Point 1, Point 2, Point 5
row 3: Disarm XY, , Back

radry commented:
@digant73 
> Main menu:
row 1: Heat/Fan, Movement, Un(Load)
row 2: Terminal, Custom, Settings
row 3: Em. Stop, , Back

> Move menu:
row 1: Z-, Y+, Z+
row 2: X-, Y-, X+
row 3: 0.1mm, , Back

> Level menu:
row 1: Point 4, Point 3, Distance
row 2: Point 1, Point 2, Point 5
row 3: Disarm XY, , Back

That's already how it is, look at the screenshots.  

> Fort the Printing menu, please provide a screenshot (and possibly consider to apply the changes in my first post).

working on it

>Also, the portrait mode is an alternative to the standard mode, so I would convert all the "#ifdef PORTRAIT...#else...#endif" to "#ifndef PORTRAIT...#else...#endif"

The code for "TFT70_V3_0" is also in there with `#ifdef TFT70_V3_0` instead of `#ifndef` although it's not the default screen. I think it's easier to read and understand with `#ifdef` and it follows the already exising coding standard of TFT70_V3_0.

radry commented:
New printing screen  

![printing_new](https://user-images.githubusercontent.com/8878118/144763575-8598dc6a-6b34-418b-b326-c23b1b7ef20d.jpg)

digant73 commented:
@radry I would update the pictures in the main section with the final layouts.
In the Printing menu I would also move the progress bar just above the Pause/More/Stop buttons just to separate the live icons / stats from the standard control buttons

radry commented:
Progress Bar on top now. 
Looks better than dividing the icons and easier to implement than having it between bottom row and Babystep button.
![printing_new_1](https://user-images.githubusercontent.com/8878118/144910010-220d159c-b2c3-43cd-a246-27171a65f000.jpg)

guruathwal commented:
@radry Nice work., but I am afraid to say that this implementation will be incompatible with the new menu API. Hardcoded menu layouts are being replaced with an easy-to-use grid system for all standard menus in my upcoming PRs.

radry commented:
@guruathwal 
Then you should implement the vertical GUI in the new menu API as well. 
I don't know how that new API works but since the hardcoded vertical layout follows the same rules and coding style as the landscape one, it should be easy to translate to your new API. I might be able to help.

digant73 commented on TFT/src/User/Menu/StatusScreen.c:
```diff
@@ -340,7 +340,7 @@ void drawStatus(void)
   GUI_SetColor(GANTRYLBL_COLOR);
   GUI_SetBkColor(infoSettings.status_xyz_bg_color);
-  sprintf(tempstr, XYZ_STATUS, coordinateGetAxisActual(X_AXIS), coordinateGetAxisActual(Y_AXIS),
+  sprintf(tempstr, XYZ_STATUS, coordinateGetAxisActual(X_AXIS), coordinateGetAxisActual(Y_AXIS), 
```
remove the trailing space in this line

radry commented:
Build Test Fails because #2281 broke this PR. I already resolved the conflicts but haven't added #2281 to my fork.

@bigtreetech Please let me know what else you require to merge this PR.

digant73 commented on TFT/src/User/Menu/Move.c:
```diff
@@ -28,17 +28,23 @@ void storeMoveCmd(AXIS xyz, int8_t direction)
 void drawXYZ(void)
 {
-  char tempstr[20];
+  char tempstr[20]; 
```
remove trailing space

digant73 commented on TFT/src/User/Menu/StatusScreen.c:
```diff
@@ -323,13 +337,12 @@ void drawStatus(void)
   #endif
 
-  GUI_SetTextMode(GUI_TEXTMODE_NORMAL);
+  GUI_SetTextMode(GUI_TEXTMODE_NORMAL);  
```
remove trailing space

digant73 commented on TFT/src/User/API/menu.c:
```diff
@@ -3,14 +3,52 @@
 #include "ListItem.h"
 #include "Notification.h"
 
+
```
remove this empty line

digant73 commented on TFT/src/User/API/menu.c:
```diff
@@ -448,7 +709,7 @@ void busyIndicator(SYS_STATUS status)
   if (status == STATUS_BUSY)
   {
-    GUI_SetColor(MENU_BUSY_DOT_COLOR);
+    GUI_SetColor(YELLOW);
```
restore GUI_SetColor(MENU_BUSY_DOT_COLOR);

digant73 commented on TFT/src/User/Menu/Move.c:
```diff
@@ -44,7 +45,7 @@ void drawXYZ(void)
     GUI_DispString(START_X + 3 * SPACE_X + 3 * ICON_WIDTH, (ICON_START_Y - BYTE_HEIGHT) / 2, (uint8_t *)tempstr);
   #endif
-
+  
```
remove trailing space

digant73 commented on TFT/src/User/Menu/StatusScreen.c:
```diff
@@ -337,13 +337,12 @@ void drawStatus(void)
-  GUI_SetTextMode(GUI_TEXTMODE_NORMAL);
-  GUI_SetColor(GANTRYLBL_COLOR);
-  GUI_SetBkColor(infoSettings.status_xyz_bg_color);
+  GUI_SetTextMode(GUI_TEXTMODE_NORMAL);  
```
remove trailing space
